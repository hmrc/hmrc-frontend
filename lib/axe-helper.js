const { configureAxe } = require('jest-axe');

const axe = configureAxe({
  rules: {
    // As we're testing incomplete HTML fragments, we don't expect there to be a
    // skip link.
    'skip-link': { enabled: false },
  },
});

module.exports = axe;
