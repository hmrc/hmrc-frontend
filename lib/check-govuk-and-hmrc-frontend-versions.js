const fs = require('fs');
const path = require('path');

const projectDir = process.env.INIT_CWD;
const hmrcFrontendDir = path.join(projectDir, 'node_modules/hmrc-frontend/');
let govukFrontendVersion = '';

try {
  govukFrontendVersion = JSON.parse(fs.readFileSync(path.join(hmrcFrontendDir, 'package.json'), 'utf8')).dependencies['govuk-frontend'].replace(/[^0-9.]/g, '');
} catch (err) {
  process.exit(0); // Exit if no package.json is found under project_path/node_modules/hmrc-frontend
}

const govukPrototypeKitPackagePath = path.join(projectDir, 'package.json');
const govukPrototypeKitConfig = JSON.parse(fs.readFileSync(govukPrototypeKitPackagePath, 'utf8'));
const currentGovukFrontendVersion = govukPrototypeKitConfig.dependencies['govuk-frontend'].replace(/[^0-9.]/g, '');
if (currentGovukFrontendVersion >= govukFrontendVersion) {
  process.exit(0);
} else {
  govukPrototypeKitConfig.dependencies['govuk-frontend'] = govukFrontendVersion;
  // eslint-disable-next-line no-console
  console.warn(`
  Hello friend, looks like you're on
  govuk-frontend v${currentGovukFrontendVersion}, 
  but hmrc-frontend requires at least v${govukFrontendVersion}.
  This error is just a heads up that hmrc-frontend might not
  work as intended on your current version.
    - PlatUI
  -------------------------------------------
          \\   ^__^ 
           \\  (oo)_______
              (__)       )\\/\\
                  ||----w |
                  ||     ||
  - PlatMooI the PlatUI Helper Cow
  `);

  process.exit(1);
}
