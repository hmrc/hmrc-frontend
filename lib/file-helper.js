const fs = require('fs');
const path = require('path');
const yaml = require('js-yaml');
const Remarkable = require('remarkable');

const configPaths = require('../config/paths.json');

const childDirectories = (dir) => fs.readdirSync(dir)
  .filter((file) => {
    const filePath = path.join(configPaths.components, file);
    const isDirectory = fs.statSync(filePath).isDirectory();
    const hasComponentConfig = fs.existsSync(`${filePath}/${file}.yaml`);
    return isDirectory && hasComponentConfig;
  });

// Generate component list from source directory, excluding anything that's not
// a directory (for example, .DS_Store files)
exports.allComponents = childDirectories(configPaths.components);

// Read the contents of a file from a given path
const readFileContents = (filePath) => fs.readFileSync(filePath, 'utf8');

exports.readFileContents = readFileContents;

const getComponentData = (componentName) => {
  const yamlPath = path.join(configPaths.components, componentName, `${componentName}.yaml`);
  return yaml.safeLoad(fs.readFileSync(yamlPath, 'utf8'), { json: true });
};

exports.getComponentData = getComponentData;

const md = new Remarkable();

const govukify = (html) => html
  .replace(/<p>/g, '<p class="govuk-body">')
  .replace(/<h1/g, '<h1 class="govuk-heading-l"')
  .replace(/<h2/g, '<h2 class="govuk-heading-m"')
  .replace(/<h3/g, '<h3 class="govuk-heading-s"')
  .replace(/<ul/g, '<ul class="govuk-list govuk-list--bullet govuk-list--spaced"')
  .replace(/<a /g, '<a class="govuk-link" ');

const getComponentReadme = (componentPath) => {
  const readme = path.join(configPaths.components, componentPath, 'README.md');
  const markdown = readFileContents(readme);
  const html = md.render(markdown);
  return govukify(html);
};

exports.getComponentReadme = getComponentReadme;
