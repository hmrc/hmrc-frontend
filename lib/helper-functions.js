// Convert component name to macro name
//
// This helper function takes a component name and returns the corresponding
// macro name.
//
// Component names are lowercase, dash-separated strings (button, date-input),
// whilst macro names have a `govuk` prefix and are camel cased (govukButton,
// govukDateInput).
const componentNameToMacroName = (componentName) => {
  const macroName = componentName
    .toLowerCase()
    .split('-')
    // capitalize each 'word'
    .map((word) => word.charAt(0).toUpperCase() + word.slice(1))
    .join('');

  return `hmrc${macroName}`;
};
exports.componentNameToMacroName = componentNameToMacroName;

exports.componentNameToComponentDirectory = (componentName) => [componentName].join('-'
  + '');
