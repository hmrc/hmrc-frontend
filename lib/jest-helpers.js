const path = require('path');
const fs = require('fs');
const nunjucks = require('nunjucks');
const cheerio = require('cheerio');
const yaml = require('js-yaml');
require('jest-axe/extend-expect');

const configPaths = require('../config/paths.json');
const { componentNameToMacroName } = require('./helper-functions');

nunjucks.configure(
  [
    configPaths.components,
    path.join(configPaths.govukFrontend, 'dist'),
    path.join(configPaths.govukFrontend, 'dist/govuk/components'),
  ],
  { trimBlocks: true, lstripBlocks: true },
);

/**
 * Render a component's macro for testing
 * @param {string} componentName
 * @param {string} params parameters that are used in the component macro
 * @param {any} children any child components or text, pass the children to the macro
 * @returns {function} returns cheerio (jQuery) instance of the macro for easy DOM querying
 */
function render(componentName, params, children = false) {
  if (typeof params === 'undefined') {
    throw new Error('Parameters passed to `render` should be an object but are undefined');
  }

  const macroName = componentNameToMacroName(componentName);
  const macroParams = JSON.stringify(params, null, 2);

  let macroString = `{%- from "${componentName}/macro.njk" import ${macroName} -%}`;

  // If we're nesting child components or text, pass the children to the macro
  // using the 'caller' Nunjucks feature
  if (children) {
    macroString += `{%- call ${macroName}(${macroParams}) -%}${children}{%- endcall -%}`;
  } else {
    macroString += `{{- ${macroName}(${macroParams}) -}}`;
  }

  const output = nunjucks.renderString(macroString);

  return cheerio.load(output);
}

/**
 * Get examples from a component's metadata file
 * @param {string} componentName
 * @returns {object} returns object that includes all examples at once
 */
function getExamples(componentName) {
  const file = fs.readFileSync(
    path.join(configPaths.components, componentName, `${componentName}.yaml`),
    'utf8',
  );

  const docs = yaml.safeLoad(file);

  const examples = {};

  docs.examples.forEach((example) => {
    examples[example.name] = example.data;
  });

  return examples;
}

/**
 * Get the raw HTML representation of a component, and remove any other
 * child elements that do not match the component.
 * Relies on B.E.M naming ensuring that child components relating to a component
 * are namespaced.
 * @param {function} $ requires an instance of cheerio (jQuery) that includes the
 * rendered component.
 * @param {string} className the top level class 'Block' in B.E.M terminology
 * @returns {string} returns HTML
 */
function htmlWithClassName($, className) {
  const $component = $(className);
  const classSelector = className.replace('.', '');
  // Remove all other elements that do not match this component
  $component.find(`[class]:not([class^=${classSelector}])`).remove();
  return $.html($component);
}

module.exports = { render, getExamples, htmlWithClassName };
