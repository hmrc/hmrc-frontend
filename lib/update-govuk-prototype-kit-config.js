const path = require('path');
const fs = require('fs');

const projectDir = path.dirname(__dirname);
const govukFrontendVersion = JSON.parse(fs.readFileSync(path.join(projectDir, 'package.json'), 'utf8')).dependencies['govuk-frontend'].replace(/[^0-9.]/g, '');
const govukPrototypeKitConfigPath = path.join(projectDir, 'src', 'govuk-prototype-kit.config.json');
const govukPrototypeKitConfig = JSON.parse(fs.readFileSync(govukPrototypeKitConfigPath, 'utf8'));
govukPrototypeKitConfig.pluginDependencies = [
  {
    packageName: 'govuk-frontend',
    minVersion: govukFrontendVersion,
  },
];
fs.writeFileSync(govukPrototypeKitConfigPath, JSON.stringify(govukPrototypeKitConfig, null, 2));
