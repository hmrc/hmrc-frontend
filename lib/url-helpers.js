import { ports } from '../config/paths.json';

export function examplePreview(example) {
  return `http://localhost:${ports.app}/components/${example.replace(/^\/+|\/+$/g, '')}/preview`;
}
