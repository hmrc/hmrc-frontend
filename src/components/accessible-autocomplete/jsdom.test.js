/* eslint-env jest */

import axe from '../../../lib/axe-helper';
import { render, getExamples, htmlWithClassName } from '../../../lib/jest-helpers';

const examples = getExamples('accessible-autocomplete');

describe('Accessible autocomplete', () => {
  describe('by default', () => {
    it('passes accessibility tests', async () => {
      const $ = render('accessible-autocomplete', examples.default);

      const results = await axe($.html(), {
        rules: {
          region: { enabled: false },
        },
      });
      expect(results)
        .toHaveNoViolations();
    });

    it('renders with id', () => {
      const $ = render('accessible-autocomplete', examples.default);

      const $component = $('select');
      expect($component.attr('id')).toEqual('location-picker');
    });

    it('renders with all required data attributes', () => {
      const $ = render('accessible-autocomplete', examples.default);

      const $component = $('select');
      expect($component.attr('data-module')).toEqual('hmrc-accessible-autocomplete');
      expect($component.attr('data-default-value')).toEqual('');
      expect($component.attr('data-auto-select')).toEqual('false');
      expect($component.attr('data-show-all-values')).toEqual('false');
      expect($component.attr('data-show-all-values')).toBeTruthy();
    });

    it('renders with a form group wrapper', () => {
      const $ = render('accessible-autocomplete', examples.default);

      const $formGroup = $('.govuk-form-group');
      expect($formGroup.length).toBeTruthy();
    });
  });

  describe('when it includes a hint', () => {
    it('renders the hint', () => {
      const $ = render('accessible-autocomplete', examples['with-label-and-hint']);

      expect(htmlWithClassName($, '.govuk-hint')).toMatchSnapshot();
    });
  });

  describe('when it includes an error message', () => {
    it('renders the error message', () => {
      const $ = render('accessible-autocomplete', examples['with-label-and-hint-and-error']);

      expect(htmlWithClassName($, '.govuk-error-message')).toMatchSnapshot();
    });

    it('renders with a form group wrapper that has an error state', () => {
      const $ = render('accessible-autocomplete', examples['with-label-and-hint-and-error']);

      const $formGroup = $('.govuk-form-group');
      expect($formGroup.hasClass('govuk-form-group--error')).toBeTruthy();
    });
  });

  describe('with dependant components', () => {
    it('renders with label', () => {
      const $ = render('accessible-autocomplete', examples['with-label-and-hint']);

      expect(htmlWithClassName($, '.govuk-label')).toMatchSnapshot();
    });
  });
});
