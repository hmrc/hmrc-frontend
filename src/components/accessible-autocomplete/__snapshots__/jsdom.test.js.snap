// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accessible autocomplete when it includes a hint renders the hint 1`] = `

<div id="location-picker-hint"
     class="govuk-hint"
>
  Please select a country
</div>

`;

exports[`Accessible autocomplete when it includes an error message renders the error message 1`] = `

<p id="location-picker-error"
   class="govuk-error-message"
>
</p>

`;

exports[`Accessible autocomplete with dependant components renders with label 1`] = `

<label class="govuk-label"
       for="location-picker"
>
  Select country
</label>

`;
